lib/trento/application/integration/discovery/payloads/sap_system_discovery_payload.ex

```diff
@@ -0,0 +1,349 @@
+      _ -> nil
+    end)
+  end
+
+  def find_property(_, _), do: nil
+end
+
+defmodule Trento.Integration.Discovery.SapSystemDiscoveryPayload.SapControlProperty do
+  @moduledoc """
+  SAP control property field payload
+  """
+
+  @required_fields [:value, :property]
+
+  use Trento.Type
+
+  deftype do
+    field :value, :string
+    field :property, :string
+    field :propertytype, :string
+  end
+
+  def changeset(property, attrs) do
+    property
+    |> cast(attrs, fields())
+    |> validate_required_fields(@required_fields)
+  end
+end
+
+defmodule Trento.Integration.Discovery.SapSystemDiscoveryPayload.SapControlInstance do
+  @moduledoc """
+  SAP control instances field payload
+  """
+
+  @required_fields [
+    :features,
+    :hostname,
+    :httpPort,
+    :httpsPort,
+    :dispstatus,
+    :instanceNr,
+    :startPriority
+  ]
+
+  use Trento.Type
+
+  deftype do
+    # current_instance is a custom field to make data extraction easier
+    field :current_instance, :boolean, default: false
+    field :features, :string
+    field :hostname, :string
+    field :httpPort, :integer
+    field :httpsPort, :integer
+
+    field :dispstatus, Ecto.Enum,
+      values: [
+        :"SAPControl-GREEN",
+        :"SAPControl-YELLOW",
+        :"SAPControl-RED",
+        :"SAPControl-GRAY"
+      ]
+
+    field :instanceNr, :integer
+    field :startPriority, :string
+  end
+
+  def changeset(instance, attrs, hostname, instance_number) do
+    enriched_attrs = enrich_current_instance(attrs, hostname, instance_number)
+
+    instance
+    |> cast(enriched_attrs, fields())
+    |> validate_required_fields(@required_fields)
+  end
+
+  defp enrich_current_instance(
+         %{"hostname" => current_hostname, "instanceNr" => current_instance_number} = attrs,
+         hostname,
+         instance_number
+       )
+       when hostname == current_hostname and instance_number == current_instance_number,
+       do: Map.put(attrs, "current_instance", true)
+
+  defp enrich_current_instance(attrs, _, _), do: attrs
+end
+
+defmodule Trento.Integration.Discovery.SapSystemDiscoveryPayload.SapControlProcess do
+  @moduledoc """
+  SAP control process field payload
+  """
+
+  @required_fields [
+    :pid,
+    :name,
+    :starttime,
+    :dispstatus,
+    :textstatus,
+    :description,
+    :elapsedtime
+  ]
+
+  use Trento.Type
+
+  deftype do
+    field :pid, :integer
+    field :name, :string
+    field :starttime, :string
+
+    field :dispstatus, Ecto.Enum,
+      values: [
+        :"SAPControl-GREEN",
+        :"SAPControl-YELLOW",
+        :"SAPControl-RED",
+        :"SAPControl-GRAY"
+      ]
+
+    field :description, :string
+    field :elapsedtime, :string
+  end
+
+  def changeset(process, attrs) do
+    process
+    |> cast(attrs, fields())
+    |> validate_required_fields(@required_fields)
+  end
+end
+
+defmodule Trento.Integration.Discovery.SapSystemDiscoveryPayload.SystemReplication do
```

Comment:
I don't know if you like it this way but you can avoid writing the whole string and define just `SystemReplication` inside the `Trento.Integration.Discovery.SapSystemDiscoveryPayload` module, will be the same

Reply:
Do you mean embedding the these new modules withing the 1st `defmodule`?

Reply:
Yes! :smile:

Reply:
@dottorblaster Modified here: https://github.com/trento-project/web/pull/602/commits/f895dbf95039ed84279002600363a26fa0f9ecea